codalab/lib/bundle_store.py

```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+            print >> sys.stderr, "cleanup: data %s" % absolute_path
             if not dry_run:
                 path_util.remove(absolute_path)
 
     def full_cleanup(self, model, dry_run):
-        '''
+        """
         For each data hash in the store, check if it should be garbage collected and
         delete its data if so. In addition, delete any old temporary files.
-        '''
+        """
         old_data_files = self.list_old_files(self.data, self.DATA_CLEANUP_TIME)
         for data_hash in old_data_files:
             self.cleanup(model, data_hash, [], dry_run)
         old_temp_files = self.list_old_files(self.temp, self.TEMP_CLEANUP_TIME)
         for temp_file in old_temp_files:
             temp_path = os.path.join(self.temp, temp_file)
-            print >>sys.stderr, "cleanup: temp %s" % temp_path
+            print >> sys.stderr, "cleanup: temp %s" % temp_path
             if not dry_run:
                 path_util.remove(temp_path)
 
     def list_old_files(self, path, cleanup_time):
+        """
+        Returns a list of old files that have not been modified since `cleanup_time` seconds ago.
+        """
         cleanup_cutoff = time.time() - cleanup_time
         result = []
-        for file in os.listdir(path):
-            absolute_path = os.path.join(path, file)
+        for entry in os.listdir(path):
+            absolute_path = os.path.join(path, entry)
             if path_util.getmtime(absolute_path) < cleanup_cutoff:
-                result.append(file)
+                result.append(entry)
         return result
+
+
+class MultiDiskBundleStore(BaseBundleStore, BundleStoreCleanupMixin):
+    """
+    A BundleMultiStore is responsible for taking a set of locations and load-balancing the placement of
+    bundle data between the locations. There are a few caveats: this is not the perfect method of
+    distributing load, as it works on a per-bundle basis, the fullness of directories is not taken into account.
+
+    Questions that this needs to answer:
+    Q: Where is the data for a bundle placed?
+    A: The directory a bundle's data is placed into is determined by the hash ring, based on its uuid. Similarly,
+       when getting the directory for a bundle, a lookup is performed by the bundle UUID.
+
+    Q: What happens if a disk fails?
+    A: No plans for replication right now, but we can think about this later. In fact, replicating the bundle data is
+       probably something that makes sense long-term.
+
+    Q: What happens if we want to migrate away from one disk to another.
+    A: The hash ring takes care of this, and only the bundles on the disk we are migrating from will be moved.
+
+    Q: What happens when the disk a bundle is supposed to go to runs out of space?
+    A: Need to avoid this scenario by proactively adding new disk mounts when the time arises, and then
+       redistributing bundles when they need to be redistributed.
+    """
```

Comment:
Write a comment here that if we have disks D1 and D2 with sizes 3N and N, then one should add 3 logical disks under D1.  Basically users have to manage their own disks.